File: backend/app/routers/resources.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.services.resource_service import ResourceService

router = APIRouter(prefix="/resources", tags=["resources"])


@router.get("/videos")
def list_videos(
    keyword: str | None = None,
    category: str | None = None,
    source: str | None = None,
    topicId: int | None = None,
    favorite: bool | None = None,
    sort: str = Query(default="latest", pattern="^(latest|hot|favorite)$"),
    db: Session = Depends(get_db),
) -> dict:
    return ResourceService(db).list_videos(keyword, category, source, topicId, favorite, sort)


@router.post("/videos")
def create_video(payload: dict, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).create_video(payload)


@router.put("/videos/{video_id}")
def update_video(video_id: int, payload: dict, db: Session = Depends(get_db)) -> dict:
    video = ResourceService(db).update_video(video_id, payload)
    if video is None:
        raise HTTPException(status_code=404, detail="Video not found.")
    return video


@router.delete("/videos/{video_id}")
def delete_video(video_id: int, db: Session = Depends(get_db)) -> dict:
    return {"deleted": ResourceService(db).delete_video(video_id)}


@router.post("/videos/{video_id}/favorite")
def set_video_favorite(video_id: int, payload: dict, db: Session = Depends(get_db)) -> dict:
    video = ResourceService(db).set_favorite(video_id, bool(payload.get("favorite", True)))
    if video is None:
        raise HTTPException(status_code=404, detail="Video not found.")
    return video


@router.patch("/videos/{video_id}/notes")
def save_video_notes(video_id: int, payload: dict, db: Session = Depends(get_db)) -> dict:
    video = ResourceService(db).save_notes(video_id, str(payload.get("notes", "")))
    if video is None:
        raise HTTPException(status_code=404, detail="Video not found.")
    return video


@router.get("/categories")
def list_categories(db: Session = Depends(get_db)) -> list[dict]:
    return ResourceService(db).list_categories()


@router.post("/categories")
def create_category(payload: dict, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).create_category(payload)


@router.put("/categories/{category_id}")
def update_category(category_id: int, payload: dict, db: Session = Depends(get_db)) -> dict:
    category = ResourceService(db).update_category(category_id, payload)
    if category is None:
        raise HTTPException(status_code=404, detail="Category not found.")
    return category


@router.delete("/categories/{category_id}")
def delete_category(category_id: int, db: Session = Depends(get_db)) -> dict:
    return {"deleted": ResourceService(db).delete_category(category_id)}


@router.get("/categories/tree")
def category_tree(db: Session = Depends(get_db)) -> list[dict]:
    return ResourceService(db).category_tree()


@router.patch("/categories/{category_id}/status")
def set_category_status(category_id: int, payload: dict, db: Session = Depends(get_db)) -> dict:
    category = ResourceService(db).update_category(category_id, {"status": payload.get("status", "启用")})
    if category is None:
        raise HTTPException(status_code=404, detail="Category not found.")
    return category


@router.post("/categories/bulk-import")
def bulk_import_categories(payload: dict, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).bulk_import_categories(payload.get("items", []))


@router.get("/categories/export")
def export_categories(db: Session = Depends(get_db)) -> dict:
    items = ResourceService(db).list_categories()
    return {"items": items, "total": len(items)}


@router.get("/topics")
def list_topics(db: Session = Depends(get_db)) -> list[dict]:
    return ResourceService(db).list_topics()


@router.post("/topics")
def create_topic(payload: dict, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).create_topic(payload)


@router.get("/topics/{topic_id}/videos")
def topic_videos(topic_id: int, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).topic_videos(topic_id)


@router.post("/topics/{topic_id}/videos")
def add_video_to_topic(topic_id: int, payload: dict, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).add_video_to_topic(topic_id, int(payload["videoId"]))


@router.delete("/topics/{topic_id}/videos/{video_id}")
def remove_video_from_topic(topic_id: int, video_id: int, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).remove_video_from_topic(topic_id, video_id)


@router.get("/notifications")
def list_notifications(db: Session = Depends(get_db)) -> list[dict]:
    return ResourceService(db).list_notifications()


@router.post("/notifications")
def create_notification(payload: dict, db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).create_notification(payload)


@router.patch("/notifications/{notification_id}/read")
def mark_notification_read(notification_id: int, db: Session = Depends(get_db)) -> dict:
    notification = ResourceService(db).mark_notification_read(notification_id)
    if notification is None:
        raise HTTPException(status_code=404, detail="Notification not found.")
    return notification


@router.post("/notifications/read-all")
def mark_all_notifications_read(db: Session = Depends(get_db)) -> dict:
    return ResourceService(db).mark_all_notifications_read()

File: frontend/src/api/resources.ts
import { apiClient } from "./client";

export interface ResourceVideoInfo {
  taskId: number;
  url: string;
  finalUrl?: string;
  platform?: string;
  title?: string;
  author?: string;
  description?: string;
  thumbnailUrl?: string;
  duration?: string;
  publishedAt?: string;
  viewCount?: string;
  directVideoUrl?: string;
  downloadRequested: boolean;
  downloadStatus: "not_requested" | "pending" | "completed" | "blocked" | "failed";
  downloadUrl?: string;
  storagePath?: string;
  status: "pending" | "fetching" | "completed" | "blocked" | "failed";
  errorMessage?: string;
}

export interface ResourceVideo {
  id: number;
  title: string;
  source: "Bilibili" | "YouTube" | "官网" | "高校公开课" | "项目组" | string;
  sourceUrl: string;
  finalUrl?: string;
  platform?: string;
  category: string;
  tags: string[];
  description: string;
  cover: string;
  thumbnailUrl?: string;
  duration: string;
  views: string | number;
  directVideoUrl?: string;
  downloadRequested: boolean;
  downloadStatus: "not_requested" | "pending" | "completed" | "blocked" | "failed";
  downloadUrl?: string;
  playUrl: string;
  videoTaskId?: number;
  videoInfo?: ResourceVideoInfo;
  favorite: boolean;
  createdAt: string;
  notes?: string;
}

export interface ResourceVideoList {
  items: ResourceVideo[];
  total: number;
}

export interface ResourceCategory {
  id: number;
  name: string;
  parent: string | null;
  icon: string;
  order: number;
  status: "启用" | "停用" | string;
  description: string;
  updatedAt: string;
  children?: ResourceCategory[];
}

export interface ResourceTopic {
  id: number;
  title: string;
  description: string;
  count: number;
  accent?: "green" | "blue" | "cyan" | "amber" | "lime" | string;
}

export interface ResourceNotification {
  id: number;
  title: string;
  message: string;
  time: string;
  read: boolean;
}

export interface ResourceVideoQuery {
  keyword?: string;
  category?: string;
  source?: string;
  topicId?: number;
  favorite?: boolean;
  sort?: "latest" | "hot" | "favorite";
}

export interface CreateResourceVideoInput {
  videoTaskId?: number;
  title?: string;
  source?: string;
  sourceUrl?: string;
  category?: string;
  tags?: string[];
  description?: string;
  cover?: string;
  duration?: string;
  views?: string | number;
  favorite?: boolean;
  notes?: string;
}

export interface UpsertResourceCategoryInput {
  name: string;
  parent?: string | null;
  icon?: string;
  order?: number;
  status?: "启用" | "停用" | string;
  description?: string;
}

export interface CreateResourceTopicInput {
  title: string;
  description?: string;
  accent?: "green" | "blue" | "cyan" | "amber" | "lime" | string;
}

export interface CreateResourceNotificationInput {
  title: string;
  message?: string;
}

export interface TopicVideoAddResponse {
  topicId: number;
  videoId: number;
  added: boolean;
  video: ResourceVideo;
}

export function fetchResourceVideos(params: ResourceVideoQuery = {}) {
  return apiClient.get<ResourceVideoList>("/resources/videos", { params });
}

export function createResourceVideo(payload: CreateResourceVideoInput) {
  return apiClient.post<ResourceVideo>("/resources/videos", payload);
}

export function updateResourceVideo(videoId: number, payload: Partial<CreateResourceVideoInput>) {
  return apiClient.put<ResourceVideo>(`/resources/videos/${videoId}`, payload);
}

export function deleteResourceVideo(videoId: number) {
  return apiClient.delete<{ deleted: boolean }>(`/resources/videos/${videoId}`);
}

export function setResourceVideoFavorite(videoId: number, favorite: boolean) {
  return apiClient.post<ResourceVideo>(`/resources/videos/${videoId}/favorite`, { favorite });
}

export function saveResourceVideoNotes(videoId: number, notes: string) {
  return apiClient.patch<ResourceVideo>(`/resources/videos/${videoId}/notes`, { notes });
}

export function fetchResourceCategories() {
  return apiClient.get<ResourceCategory[]>("/resources/categories");
}

export function createResourceCategory(payload: UpsertResourceCategoryInput) {
  return apiClient.post<ResourceCategory>("/resources/categories", payload);
}

export function updateResourceCategory(categoryId: number, payload: Partial<UpsertResourceCategoryInput>) {
  return apiClient.put<ResourceCategory>(`/resources/categories/${categoryId}`, payload);
}

export function deleteResourceCategory(categoryId: number) {
  return apiClient.delete<{ deleted: boolean }>(`/resources/categories/${categoryId}`);
}

export function fetchResourceCategoryTree() {
  return apiClient.get<ResourceCategory[]>("/resources/categories/tree");
}

export function setResourceCategoryStatus(categoryId: number, status: "启用" | "停用" | string) {
  return apiClient.patch<ResourceCategory>(`/resources/categories/${categoryId}/status`, { status });
}

export function bulkImportResourceCategories(items: UpsertResourceCategoryInput[]) {
  return apiClient.post<{ created: number; items: ResourceCategory[] }>("/resources/categories/bulk-import", { items });
}

export function exportResourceCategories() {
  return apiClient.get<{ items: ResourceCategory[]; total: number }>("/resources/categories/export");
}

export function fetchResourceTopics() {
  return apiClient.get<ResourceTopic[]>("/resources/topics");
}

export function createResourceTopic(payload: CreateResourceTopicInput) {
  return apiClient.post<ResourceTopic>("/resources/topics", payload);
}

export function fetchResourceTopicVideos(topicId: number) {
  return apiClient.get<ResourceVideoList>(`/resources/topics/${topicId}/videos`);
}

export function addResourceVideoToTopic(topicId: number, videoId: number) {
  return apiClient.post<TopicVideoAddResponse>(`/resources/topics/${topicId}/videos`, { videoId });
}

export function removeResourceVideoFromTopic(topicId: number, videoId: number) {
  return apiClient.delete<{ topicId: number; videoId: number; removed: boolean }>(`/resources/topics/${topicId}/videos/${videoId}`);
}

export function fetchResourceNotifications() {
  return apiClient.get<ResourceNotification[]>("/resources/notifications");
}

export function createResourceNotification(payload: CreateResourceNotificationInput) {
  return apiClient.post<ResourceNotification>("/resources/notifications", payload);
}

export function markResourceNotificationRead(notificationId: number) {
  return apiClient.patch<ResourceNotification>(`/resources/notifications/${notificationId}/read`);
}

export function markAllResourceNotificationsRead() {
  return apiClient.post<{ updated: number }>("/resources/notifications/read-all");
}

export function resourceVideoPlayableUrl(video: ResourceVideo) {
  return video.downloadUrl ?? video.directVideoUrl ?? video.playUrl ?? video.sourceUrl;
}
